### DISCUSSION TITLE
Should I be able to copy and paste from/to the terminal?

### DISCUSSION THREAD

**Discussion by LorenAmelang** [holzschu/a-shell#704] [Q&A]
I apologize if this should be obvious, but it isn't to me. Maybe because my iPad has an Apple Magic Mouse and a USB keyboard that was designed for Windows...  And it is in dark mode... 

I can drag select characters and the background changes and stays changed. It looks like I can drag down and select multiple lines. But then what? 

I've seen a black popup with "Paste" appear when I right click the mouse, but not always, and nothing has ever been pasted. 
When I select and right click, I get a similar popup, but it is smaller, empty, solid black. 

There are very pale, as if disabled, icons at the bottom of the screen, right of the keyboard icon - they never do anything. But the four direction arrows right of those icons move the cursor if I click them. Maybe I don't understand the required sequence? 

What am I missing?

**Comment by holzschu** [comment]
These are all very good questions, and I'm not sure I understand everything.
If you're using a-Shell with an external keyboard, the easiest way to copy and paste is through keyboard shortcuts (command-C and command-V, or something similar with your keyboard).

They keys at the bottom of the screen are tab, control, escape and paste (to the left) and the four arrows to the right. In my case, they appear in blue-on-black in dark mode, not "very pale, as if disabled". In the Settings app, you have options to control the aspect of the toolbar: "Use the iPadOS toolbar" (I recommend "On") and "Toolbar color" (you can try several settings).

Paste should work (it works on my iPad), but it only pastes on the command line, not on the screen. 

After receiving your issue, I've run several tests, and working only with the mouse is complicated. It's good for selection, or for moving the cursor, but after that copying and pasting is better done with either the keyboard or a click on the screen.

If you can, I would be interested in some screenshots of how things appear on your machine. It seems different from what I can observe, and also not how it should look like according to the developer documentation.

**Reply by LorenAmelang** [reply]
After trying many combinations, it seems the "Win" key on my USB keyboard is read as "command" by a-Shell. I haven't seen that in any other app...  Win+C and Win+V usually copy and paste as I'd expect. Except when they don't...  It is hard to tell whether failures are due to failed selection, key combinations or timing, or whether the clipboard did not contain what I assumed it did. Is there some way to view the clipboard content? Other than trying to paste it somewhere?

The toolbar keyboard and mic icons are vivid, the items between are much dimmer. I just tried disabling "Use the iPadOS toolbar", and got just the keyboard and mic at the lower right corner. Enabled it again and got no toolbar. Had to hard kill a-Shell to get back to full toolbar. The only color option my horrible vision can cope with is Dark Mode. 

The paste icon does usually work at the command prompt, if the clipboard is occupied. 

My main problem with the mouse is that the tiniest move I can make with my scroll finger moves the page content several screen heights! Just clicking the mouse often moves the content half a screen height. Luckily I usually want to return to the bottom prompt...  Some apps respond painfully slowly to the scroll, others are too fast, but a-Shell is almost uncontrollable. 

Long press Win+V menu.png shows a menu I can usually get by holding the Win key for several seconds. Quite helpful, once you know that. But for half of my testing today, nothing I tried would make it appear. After the hard kill of a-Shell, it is back. 

![Long press Win+V menu](https://github.com/holzschu/a-shell/assets/1297090/f2e956a9-66ef-471e-8227-bdd4ff776be5)

Typical Screenshot.png shows the toolbar items that look really dim to me. I have a hard time with all of Apple's dim blue option items - guess this is why they call age the "golden years"...  

![Typical Screenshot](https://github.com/holzschu/a-shell/assets/1297090/3d7defc3-cefd-4ca0-899f-a24e80732ede)

One more question - is there a way to enter home and end to jump to the ends of a long command? Those keys on my keyboard add "Ctrl+h" etc. to the command...  

Really appreciate your detailed attention!

**Comment by holzschu** [comment]
Thank you very much for the detailed explanation and the screenshots. 
I now have a better understanding of the issue. And I also know that your issue is not about a newly appeared bug: your app is behaving as it should (that is a important news for me).

I see several solutions that can help you in the short time.
- in Settings, disable "use iPadOS toolbar", then go back to a-Shell and type "showToolbar" if the toolbar is not present. This will give you white-on-black buttons (a better contrast).
- Alternatively, in Settings, enable "use iPadOS toolbar", then set "Toolbar color" to "light mode". This will give you blue-on-white buttons, not as good in terms of contrast, but perhaps better than blue-on-black.

You can also use "config" to change the text color, text size and background colors for a better readability (yellow-on-black might be better than white-on-black, but that's your choice... and it won't change the button colors).

I'm going to think about changing the button color. It's the system default, but as you point out the contrast level is not always good enough.

There is also an "Accessibility" generic part in Settings that has the option "Increase Contrast", but it doesn't seem to do much on the buttons. It also has "Colour Filters" that changes the system colors, it might increase the contrast.
